### PR TITLE
Add salmon splici indices to Refgenie

### PR DESCRIPTION
**Pull request by nadjano**
In order to use alevin-fry in the quantification pipeline for droplet snRNA and scRNA-seq we need salmon 'splici' (spliced transcripts + introns) indices.

For that we need to:

- Create assets for splici transcriptomes

- build indices from splici transcriptome with recent version of salmon

### CODE REVIEW

File: main.nf
Comment (pcm32) [inline main.nf]: I'm guessing that using fly here is temporary for some tests, correct? Make sure to change before merging.

**Review by pcm32 — COMMENTED** [review]
Looks good in general. I wonder though if we should have two separate workflows, one for the previous setup (in case we want to go back to it for whatever reason) and one for the newer splici indices.

File: main.nf
Comment (nadjano) [reply main.nf]: Oh , of course

**Comment by nadjano** [comment]
> Looks good in general. I wonder though if we should have two separate workflows, one for the previous setup (in case we want to go back to it for whatever reason) and one for the newer splici indices.

So you mean, that we might want to store the 'normal' salmon indices and the spliced ones in refgenie?

**Comment by pcm32** [comment]
I mean that maybe we should have two separate workflows or that the same workflow could be run in one mode or the other? It seems to be that in the current setup we are losing the ability to produce the normal salmon indexes, correct?

**Comment by nadjano** [comment]
Yes, it is true that in this setup we only create the splici indices and lose the normal indices. But if we switch to alevin-fry, the splici indices are also the recommended indices and the 'normal' indices cannot be used for the way alevin-fry is run.

So I don't really see the need to have the possibility to create both kinds of salmon indices, because if we change the scxa workflow to alevin-fry, we can't use the old ones anymore.

And if we keep alevin, we continue to use the 'old' workflow with the 'normal' indices.

**Comment by pcm32** [comment]
You might want to access the old ones for reproducibility, or because we find ourselves in the need of rowing back on this for whatever reason (maybe certain studies for a release are not ready or certain organisms need to be redone "the old way" for coherence). There is really no reason to lose the ability to compute the other type of indexes if having the two doesn't add too much complexity. We could rely on the version control as well, but then it will be down to someone later on to start fishing across commits to know when this change was made.

More than if, the question to me is how: is it better to have two separate pipelines or a single pipeline with some conditionality to be able to pick on runtime?

It might be down for @pmb59 to decide, as he will be the one who has to take care of this in the long term.

**Comment by pmb59** [comment]
Hi @nadjano, I agree with @pcm32 , we should keep the alevin workflow that generates normal indexes to use in case of need. This is currently implemented in production so I would not get rid of it for now.

The most elegant way would be to have a single main workflow that uses a config value (e.g. 'splici' or 'normal', for alevin-fry and alevin respectively) and run necessary processes for each. Even more elegant, you could make use of nested workflow execution and [specific workflow entry points](https://carpentries-incubator.github.io/workflows-nextflow/11-subworkflows/index.html), e.g. `$ nextflow run main.nf -entry SPLICI`, but I am not sure if this is allowed in DSL1 and/or DSL2 syntax. It might be that the `workflow{ }` was introduced for DSL2.

An alternative to that would be to keep two completely separated workflows in the repo. Maybe that's easier and avoids adding complicated control logic to a unique larger workflow. In that case, you could have them in, e.g:
https://github.com/ebi-gene-expression-group/isl_refs_to_refgenie/standard_indexes/
https://github.com/ebi-gene-expression-group/isl_refs_to_refgenie/splici_indexes/
and leave the main [README.md](https://github.com/ebi-gene-expression-group/isl_refs_to_refgenie/blob/feature/add_splici_index/README.md) with brief info on why have two independent workflows.

**Comment by nadjano** [comment]
Hi @pmb59 thanks for the suggestions! I think the nested workflows are possible in DSL1. 
I added a "mode" parameter which can be either 'normal' or 'splici' and based on this builds the respective salmon indices (https://github.com/ebi-gene-expression-group/isl_refs_to_refgenie/blob/39bb22d75f63a03d691de54b351a31dcb737ee8e/main.nf#L544).

File: README.md
Comment (pmb59) [inline README.md]: can this to the first parameter listed?

File: main.nf
Comment (pmb59) [inline main.nf:463]: Just to double check. You are removing this input (CDNA_REFERENCES_FOR_SALMON) in this process before deciding whether mode is 'normal' or 'splici'?

File: main.nf
Comment (pmb59) [inline main.nf]: could you please replace the tabs for spaces?

File: main.nf
Comment (pmb59) [inline main.nf:381]: please confirm this and other scripts in the `/bin` folder do not need any modification to allow salmon splici indexes in Refgenie.

**Review by pmb59 — APPROVED** [review]
LGTM!   @nadjano As long as you have tested that the current version can successfully produce normal and splici indexes.

Please see few comments, the most important one related to `CDNA_REFERENCE_FOR_SALMON`

File: main.nf
Comment (nadjano) [reply main.nf:463]: @pmb59 yes, that looks strange. But the input channel SPLICI_REFERENCE_FOR_SALMON is the same as CDNA_REFERENCES_FOR_SALMON so it does not make a difference, because later cDNA or splici is selected. 
To improve it, I changed it to SPLICI_REFERENCE_FOR_SALMON.join(CDNA_REFERENCE_FOR_SALMON).

File: README.md
Comment (pmb59) [inline README.md]: A small suggestion:
```suggestion
    --mode=<type of salmon index: [normal, splici]> \
```

File: main.nf
Comment (pmb59) [reply main.nf:463]: thanks @nadjano Nevertheless, if it not clear at this point if `SPLICI_REFERENCE_FOR_SALMON` will be used as normal (cDNA) or splici, better to name this generically as `REFERENCE_FOR_SALMON` and then once this is decided use appropriate prefix

**Comment by pcm32** [comment]
I discovered some issues with the current refgenie newest/current being used in production. Having played around with this I would suggest that we have an option that says 'both', so that both splici and regular are created. We will clearly need to have these co-existing for a while, and as such is better if a single refgenie assets update can include both types of indexes.

**Pull request merged**
